enhancements/monitoring/integration-with-topology-modes.md

```diff
@@ -0,0 +1,176 @@
+`SingleReplica`, it should *not* configure the operand for high-avaible
+operation.
+
+### Goals
+
+- Deploy the monitoring components with one replica for non-HA configuration.
+- Deploy the user-workload components with one replica for non-HA configuration.
+- Ensure that CMO passes the single-node tests from openshift/origin.
+
+### Non-Goals
+
+- Implement additional enhancements to reduce CPU and memory footprints.
+- Allow cluster admins to control the number of replicas per operand.
+- Address CodeReady Containers requirements (though it might help).
+- Add tests for `SingleReplica` to the existing end-to-end test suite in the CMO repository.
+
+## Proposal
+
+Whenever CMO needs to reconcile, it retrieves the
+`infrastructures.config.openshift.io/cluster` resource and reads the value of
+the `infrastructureTopology` status field. If the value is equal to
+`SingleReplica`, CMO adjusts the number of replicas to 1 for the following
+resources:
+
+* openshift-monitoring/prometheus-k8s (StatefulSet)
+* openshift-monitoring/alertmanager-main (StatefulSet)
+* openshift-monitoring/thanos-querier (Deployment)
+* openshift-monitoring/prometheus-adapter (Deployment)
+* openshift-user-workload-monitoring/prometheus-user-workload (StatefulSet)
+* openshift-user-workload-monitoring/thanos-ruler (StatefulSet)
+
+Otherwise CMO deploys the operands with their default number of replicas to
+ensure high-availability:
+
+* openshift-monitoring/prometheus-k8s (StatefulSet): 2 replicas
+* openshift-monitoring/alertmanager-main (StatefulSet): 3 replicas
+* openshift-monitoring/thanos-querier (Deployment): 2 replicas
+* openshift-monitoring/prometheus-adapter (Deployment): 2 replicas
+* openshift-user-workload-monitoring/prometheus-user-workload (StatefulSet): 2 replicas
+* openshift-user-workload-monitoring/thanos-ruler (StatefulSet): 2 replicas
+
+By default, CMO deploys no operand on control nodes thus it should not look at
+the `controlPlaneTopology` field.
+
+### Risks and Mitigations
+
+In HA mode, the Prometheus instances are spread on different nodes using
+soft anti-affinity and they work completely indepently one from another (there's no
+data replication between them). Services consuming the metrics (like the
+OpenShift console or Grafana) use the Thanos querier API which aggregates and
+deduplicates data from both Prometheus instances. This setup provides
+redundancy in term of metrics scraping, rule evaluations and querying as long
+as one of the Prometheus instances is still up.
+
+The Thanos ruler instances of user workload monitoring work the same as
+Prometheus except they only deal with rule evaluations.
+
+The 3 Alertmanager instances receive alerts from both Prometheus instances and
+as long as one instance is up and running, the notifications should flow to the
+receivers. Alertmanager instances also replicate silences and notification logs
+between themselves to prevent duplicate notifications.
```

Comment:
unrelated to this EP but why do we run 3 of them then?  2 seems sufficient for HA.

Reply:
Alertmanager is a lightweight component compared to Prometheus and with 3 replicas, we have a more resilient system even without persistent storage. More context [here](https://github.com/openshift/cluster-monitoring-operator/issues/309) but we could definitely reconsider this setting.